### PR TITLE
Merging milestone-friendshipv1 to development

### PR DESCRIPTION
**Pull request by Kyle-Law**
Good day! Friendship linkage have been set up. You can test the app by following steps:

1. clone the branch
2. bundle install
3. rails db:seed
4. sign up with your account
5. Click all user
6. when you clicked add friend, it'll show 'pending request', and you'll be able to accept or reject the invitation by going into the user page. (For the ease of experiment, the functionality be authorized later)
7. Upon accepted, it'll show 'already friend' in all users page
8. Upon rejected, 'Add friend' will be available again.

Thank you for reviewing. :)

### CODE REVIEW

File: app/models/friendship.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  after_destroy :destroy_reverse_relationship
+
+  belongs_to :user
+  belongs_to :friend, :class_name => "User"
+
+  validates_presence_of :user_id, :friend_id
+  validates :user_id, uniqueness: {  scope: :friend_id,case_sensitive: false }
+
+  # has_many :confirmed_friends, through: :friendships, source: :friend
+  # has_many :inverse_friends, through: :friendships, source: :user
```
Comment (wathigo) [inline app/models/friendship.rb]: Let's move the commented code. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from Pathwright)

File: app/models/friendship.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    update(status: true)
+    create_reverse_relationship
+  end
+  
+
+  # private
+
+  # def create_reverse_relationship
+  #   friend.friendships.create(friend: user, status: true)
+  # end
```
Comment (wathigo) [inline app/models/friendship.rb]: Let's move the commented code. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from Pathwright)

File: app/views/users/index.html.erb [hunk 3]
```diff
@@ -1,10 +1,21 @@
...
-        Name: <%= user.name %>
+        Name: 
         <span class="profile-link">
-          <%= link_to 'See Profile',  user_path(user), class: 'profile-link' %>
+          <%= link_to user.name,  user_path(user), class: 'profile-link' %>
+          <% if current_user.pending_friends.include?(user)%>
+            <h6>Pending Request</h6>
+          <% elsif current_user.friends.include?(user) %>
+            <h6>Already Friend</h6>
+          <% elsif current_user == user%>
```
Comment (wathigo) [inline app/views/users/index.html.erb]: **REQUIRED**
Let's remove all the logic from the views. Views related logic should be placed in the helpers. 
[The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

**Review by wathigo — CHANGES_REQUESTED** [review]
### Changes Required.
Great work @Kyle-Law :clap: 
Let's make some changes in order to meet all the requirements.
Currently, our application does not allow us to accept or decline a friendship. Instead, we can be able to send a friend request to a user who is a pending friend.

**REQUIRED.**
* Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from Pathwright).
* User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from Pathwright).
* User has to accept friendship request to start friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)).
* User can see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)).

**After making the changes, submit another request for review.**

**Happy Coding :computer: 
Simon Wathigo.**

**Comment by Kyle-Law** [comment]
Good day! Thanks for reviewing my project and the following refinement have been done:

- Unnecessary comments are removed
- View logics are moved to the helper
- Added Dark-Light Mode toggle
- Only friends' and user's posts are shown on the timeline
- Friendship requests can be rejected or accepted on individual user #show page

Thanks and looking forward to another review :)

* you may seed user by running `rails db:seed`

**Comment by Kyle-Law** [comment]
Good day! Thanks for reviewing my project and the following refinement have been done:

- Unnecessary comments are removed
- View logics are moved to the helper
- Added Dark-Light Mode toggle
- Only friends' and user's posts are shown on the timeline
- Friendship requests can be rejected or accepted on individual user #show page

Thanks and looking forward to another review :)

* You may check the live version here: https://social-media-kyle.herokuapp.com/ (However, the dark-light toggling will only work upon refresh, probably because the javascript file will be re-run again upon clicking different links. The dark & light state will preserve in local storage hence feel free to refresh and try the functionality.)
* The javascript works well locally but not on heroku, feel free to shed some light on that matter.

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great Progress @Kyle-Law :clap: However, some requirements need to be implemented in order to get approval for this milestone. check :arrow_down: 

- **[REQUIRED]** Some of the requirements mentioned by the previous reviewer weren't implemented.
* [ ] There is a button to invite user to friendship on users list page and on single user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
  - The invite button should hide for users that the current_user has a pending friend request with. Because otherwise, it will create duplicate friendships records and cause unexpected behaviour. In short, user shouldn't be able to send a friend request to users that he have pending friend request with.
* [ ] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

- **[REQUIRED]** Pull request should display passing  code linter checks :heavy_check_mark: 

### That's it :relieved: When finished, kindly submit a code re-review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Comment by Kyle-Law** [comment]
Thanks for reviewing. The functionality can be verified at https://social-media-kyle.herokuapp.com/

- To add friend, go to All user page, > click add friend > 'pending request' will be shown > go to the user page by click the user name > click accept > back to user page > 'Already Friend' will be shown > If rejected is selected, the request will be destroy and 'Add Friend' is possible again.
- Rubocop has been setup but doesn't show in PR. When I run 'rubocop .','31 files inspected, no offenses detected' detected is shown.
- To test, run 'rspec' ,'26 examples, 0 failures' will be shown.
- The javascript light/dark mode toggle functionality works well locally but not on heroku, feel free to shed some light on that matter.

**Review by jofer86 — APPROVED** [review]
## :wave: :wave: Hello! I'm Jorge, and  I will be conducting this code review. Let me get a closer look 🔎 🔍 🔎 🔍 🕵🏼 
- Changes marked as **[OPTIONAL]** are not hard requirements but are nice to have.
- The required changes were made! good job!! You may pass!!
## `const PROJECT_STATUS = 'Approved'` 🚀 🚀 🚀
--------------------------------------------------------------------
**Rock on**  🤘🏼 
**_Jorge Fernando Rincón Figueroa._** 
Technical Support Engineer.

**Pull request merged**